### ISSUE TITLE
WidgetManager’s removeReportWidget() Issues

### ISSUE THREAD

**Issue by DanHarrin** [octobercms/october#4524]
- OctoberCMS Build: 457
- PHP Version: 7.3.1

I’ve noticed a couple of issues with `WidgetManager`’s `removeReportWidget()` method.

1. is that the class is missing a simple `use SystemException;`, so I get `Class 'Backend\Classes\SystemException' not found` error messages thrown at me when trying to call `removeReportWidget()`.
2. is that there’s a very odd issue with the detection of whether the widgets are loaded or not.

I am using the following code in my plugin’s `boot()` method to remove the system report widgets:
``` 
Event::listen('system.reportwidgets.extendItems', function($widgetManager) {
    $widgetManager->removeReportWidget('Backend\ReportWidgets\Welcome');
    $widgetManager->removeReportWidget('Cms\ReportWidgets\ActiveTheme');
    $widgetManager->removeReportWidget('System\ReportWidgets\Status');
}); 
```

By default, this doesn’t work. It throws an `Unable to remove a widget before widgets are loaded` exception. Here is the code for the `removeReportWidget()`:
```
 /**
 * Remove a registered ReportWidget.
 * @param string $className Widget class name.
 * @return void
 */
public function removeReportWidget($className)
{
    if (!$this->reportWidgets) {
        throw new SystemException('Unable to remove a widget before widgets are loaded.');
    }

    unset($this->reportWidgets[$className]);
}
```

As you can see, the exception is thrown when `$this->reportWidgets` is just an empty array. Fine, that should work. I now edit the method’s code to see what’s going on:
``` 
dd($this->reportWidgets);

  if (!$this->reportWidgets) {
    throw new SystemException('Unable to remove a widget before widgets are loaded.');
}

unset($this->reportWidgets[$className]);
```

This code is executed, and for some reason returns an array with the three system report widgets! This shouldn’t happen, as the conditional statement’s exception suggests that `$this->reportWidgets` has no items. Odd, right? I go further.
```
 if (!$this->reportWidgets) {
    dd($this->reportWidgets);

    throw new SystemException('Unable to remove a widget before widgets are loaded.');
}

unset($this->reportWidgets[$className]);
```

And the same three-item array is returned! I’m now very confused - how is the conditional passing *and* there be three items in `$this->reportWidgets`? It truly makes no sense to me. Finally, I comment that statement out completely, just leaving the `unset()`. Sure enough, no issues. The report widgets are deregistered successfully.

I’m truly stumped about the inconsistencies here - I’m sure that I’m probably doing something wrong, and it’s an obvious fix. If someone could give me a pointer / see if they can replicate my issue, that would be great! For now though, the code is left commented while I’m developing my plugin.

**Comment by bennothommo** [comment]
@DanHarrin I've tested this, and it's not a bug per se - what's happening is the `system.reportwidgets.extendItems` event is being called twice because the [`listReportWidgets`](https://github.com/octobercms/october/blob/37eb5b0c3c535f7c0473fd9d0c02e4f405a154a6/modules/backend/classes/WidgetManager.php#L165) method in `Backend\Classes\WidgetManager` is being called twice.

Your event has removed all the widgets in the first event call, and with the second event call, it is trying to remove them again. A simple fix would be to check the `$widgetManager->reportWidgets` property for the existence of the widget *before* trying to remove it.

**Comment by DanHarrin** [comment]
@bennothommo that would work, however `$widgetManager->reportWidgets` is a protected property at the moment. I'll PR that along with the missing `SystemException` import.

**Comment by DanHarrin** [comment]
Fixed with #4525.
